Add a WithValueTranslator option to Reconciller.

porridge opened the pull request:
A Translator is a way to produces helm values based on the fetched custom
resource itself (unlike `Mapper` which can only see `Values`).

This way the code which converts the custom resource to Helm values can first
convert an `Unstructured` into a regular struct, and then rely on Go type
safety rather than work with a tree of maps from `string` to `interface{}`.

Thanks to having access to a `Context`, the code can also safely access the
network, for example in order to retrieve other resources from the k8s cluster,
when they are referenced by the custom resource.

Fixes: #113

porridge commented on pkg/values/values.go:
```diff
@@ -17,9 +17,13 @@ limitations under the License.
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
+// TODO: Consider deprecating Mapper and overrides in favour of Translator.
```
Please let me know what do you think about this. Since `Translator` is a strictly more powerful interface, perhaps we should get rid of `Mapper`?

joelanford replied:
+1 from me (pending a conversation about the timing of applying overrides)

porridge reviewed (commented):
/assign @varshaprasad96 
cc @fabianvf

varshaprasad96 commented on pkg/reconciler/reconciler.go:
```diff
@@ -534,15 +547,16 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.
 		return chartutil.Values{}, err
 	}
-	vals := r.valueMapper.Map(crVals.Map())
+	vals = r.valueMapper.Map(crVals.Map())
```
agreed, having `valueMapper` is not of much use here. Instead of using both `valueTranslator` and `valueMapper` on the same values can we prioritize the translatorfunc if its not nil by the user?

porridge replied:
@varshaprasad96 can you please clarify what you mean by "prioritize the translatorfunc if its not nil by the user"?
Do you mean passing both `WithValueMapper` and `WithValueTranslator` to `New()` should be rejected?

varshaprasad96 replied:
If `Translator` is passed, we can ignore the mapper (in case both are non-nil). This would make it easier to deprecate and remove mapper completely later. Thoughts?

porridge replied:
I'm afraid it would be detrimental to developer experience if the `WithValueMapper` option was silently ignored when the `WithValueTranslator` option would be passed. IMHO it would violate the principle of least astonishment.

varshaprasad96 replied:
Completely agree. I had logging a warning and docs along with it in mind so that we don't have to make a breaking change of removing `WithValueMapper` immediately. We can probably bring this up and discuss in tomorrow's [community meeting](https://github.com/operator-framework/community#helm-operator-plugins-working-group). It would really be nice to get this PR merged as it's a really helpful addition :)

porridge replied:
Unfortunately I'll be in a place with poor network coverage at that time so I will not be able to join the meeting. But to summarise my thoughts:
- I don't think we need to remove value mappers in this PR, or even the same release, but it would make sense to add deprecation warnings to prepare for removal if we want to do it eventually,
- a warning and docs alone are in my experience insufficient countermeasure to a surprise in the API unless there is no other option, and I think we have a good alternative:
- IF we want to make mappers and translators exclusive, then I'd strongly recommend returning and error from New in case both `WithValueMapper` and `WithValueTranslator` are passed. This is not going to harm existing users, unless they try using a translator, in which case such error is a useful hint.

porridge replied:
Hi, any updates on this after the meeting? How do we proceed on this?

Also, would it be possible to add me to the `operator-framework` org? Just so that I can configure GitHub to send PR notifications to my work email..

joelanford commented:
There's a similar PR to this one (#93).

I mostly summarized my thoughts on this subject [here](https://github.com/operator-framework/helm-operator-plugins/pull/93#issuecomment-861993865).

My main area of concern is the timing of applying value overrides:
> One notable semantic change would be that all of the translation and mapping would happen before overrides are applied. This would make setting the overrides less ergonomic since mapped values may otherwise be an internal implementation detail of the operator.

If a translator is in use, someone configuring overrides will need to be aware of how the CR is mapped to chart values because the overrides apply to the translator's output. This means potentially exposing internal chart implementation details to users applying overrides at runtime.

porridge commented:
Thanks for pointing me at the overrides documentation, @joelanford - I guess I didn't really understand its purpose before.
Here's my stab at applying overrides first. PTAL

Rebased to take advantage of #112 but otherwise the first commit should be unchanged.

joelanford commented on pkg/reconciler/reconciler.go:
```diff
@@ -375,8 +376,33 @@ func WithPostHook(h hook.PostHook) Option {
+// from a custom resource spec to the values passed to Helm.
+// Use this if you want to apply a transformation on the values obtained from your custom resource, before
+// they are passed to Helm.
+// This functionality will be removed in a future release. Use WithValueTranslator instead.
```
Start a separate paragraph with the prefix `Deprecated: `

From a Go [blog post about deprecations](https://go.dev/blog/godoc):
>  Sometimes a struct field, function, type, or even a whole package becomes redundant or unnecessary, but must be kept for compatibility with existing programs. To signal that an identifier should not be used, add a paragraph to its doc comment that begins with “Deprecated:” followed by some information about the deprecation.

Linting tools look for this and provide helpful warnings to developers using deprecated code.

```suggestion
//
// Deprecated: This functionality will be removed in a future release. Use WithValueTranslator instead.
```

porridge replied:
Done.

joelanford commented on pkg/values/values.go:
```diff
@@ -17,9 +17,13 @@ limitations under the License.
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
+// TODO: Consider deprecating Mapper in favour of Translator.
+
```
Let's go ahead and deprecate it IMO.

porridge replied:
Done.

joelanford commented on pkg/reconciler/reconciler.go:
```diff
@@ -534,15 +560,15 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.
-	crVals, err := internalvalues.FromUnstructured(obj)
-	if err != nil {
+func (r *Reconciler) getValues(ctx context.Context, obj *unstructured.Unstructured) (chartutil.Values, error) {
+	if err := internalvalues.ApplyOverrides(r.overrideValues, obj); err != nil {
```
:+1: This looks great!

porridge commented on pkg/reconciler/internal/values/values_test.go:
```diff
@@ -25,73 +26,50 @@ import (
+
+	When("Unstructured object is invalid", func() {
+		AfterEach(func() {
+			Expect(ApplyOverrides(nil, u)).NotTo(BeNil())
```
Done.

porridge commented:
Applied the suggestions and squashed the commits.
I also tried to improve the docstrings a little and make sure they look sane in godoc.
Also added a nolint pragma for the use of `Mapper` to make our lint pass.
PTAL @joelanford

joelanford reviewed (approved):
/approve
